core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -244,6 +242,88 @@ object Uri {
     }
   }
 
+  /** The userinfo subcomponent may consist of a user name and,
+    * optionally, scheme-specific information about how to gain
+    * authorization to access the resource.  The user information, if
+    * present, is followed by a commercial at-sign ("@") that delimits
+    * it from the host.
+    *
+    * @param username The username component, decoded.
+    *
+    * @param password The password, decoded.  Passing a password in
+    * clear text in a URI is a security risk and deprecated by RFC
+    * 3986, but preserved in this model for losslessness.
+    *
+    * @see https://www.ietf.org/rfc/rfc3986.txt#section-3.21.
+    */
+  final case class UserInfo private (username: String, password: Option[String])
+      extends Ordered[UserInfo] {
+    override def compare(that: UserInfo): Int =
+      username.compareTo(that.username) match {
+        case 0 => Ordering.Option[String].compare(password, that.password)
+        case cmp => cmp
+      }
+  }
+
+  object UserInfo {
+
+    /** Parses a userInfo from a percent-encoded string. */
+    def fromString(s: String): ParseResult[UserInfo] =
+      fromStringWithCharset(s, StandardCharsets.UTF_8)
+
+    /** Parses a userInfo from a string percent-encoded in a specific charset. */
+    def fromStringWithCharset(s: String, cs: NioCharset): ParseResult[UserInfo] =
+      new Http4sParser[UserInfo](s, "Invalid user info") with Rfc3986Parser {
+        def main = userInfo
+        def charset = cs
+      }.parse
+
+    private[http4s] trait Parser { self: Rfc3986Parser =>
```

Comment:
Following pattern of putting parsers in companion object, because they may call a private constructor after validation.  It does lead to some quirky self-typing.